src/lib/common/include/sol-power-supply.h

```diff
@@ -0,0 +1,360 @@
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <stdlib.h>
+
+#include <sol-vector.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file
+ * @brief Power Supply API provides a way to get information about battery,
+ * UPS, AC or DC power supply.
+ *
+ * Some information only is provided by some kinds of power supplies and
+ * depends on platform and drivers.
+ */
+
+/**
+ * @defgroup PowerSupply Power Supply
+ *
+ * @{
+ */
+
+enum sol_power_supply_type {
+    SOL_POWER_SUPPLY_TYPE_UNKNOWN,
+    SOL_POWER_SUPPLY_TYPE_BATTERY,
+    SOL_POWER_SUPPLY_TYPE_UPS,
+    SOL_POWER_SUPPLY_TYPE_MAINS,
+    SOL_POWER_SUPPLY_TYPE_USB,
+    SOL_POWER_SUPPLY_TYPE_USB_DCP,
+    SOL_POWER_SUPPLY_TYPE_USB_CDP,
+    SOL_POWER_SUPPLY_TYPE_USB_ACA
+};
+
+enum sol_power_supply_status {
+    SOL_POWER_SUPPLY_STATUS_UNKNOWN,
+    SOL_POWER_SUPPLY_STATUS_CHARGING,
+    SOL_POWER_SUPPLY_STATUS_DISCHARGING,
+    SOL_POWER_SUPPLY_STATUS_NOT_CHARGING,
+    SOL_POWER_SUPPLY_STATUS_FULL
+};
+
+enum sol_power_supply_capacity_level {
+    SOL_POWER_SUPPLY_CAPACITY_LEVEL_UNKNOWN,
+    SOL_POWER_SUPPLY_CAPACITY_LEVEL_CRITICAL,
+    SOL_POWER_SUPPLY_CAPACITY_LEVEL_LOW,
+    SOL_POWER_SUPPLY_CAPACITY_LEVEL_NORMAL,
+    SOL_POWER_SUPPLY_CAPACITY_LEVEL_HIGH,
+    SOL_POWER_SUPPLY_CAPACITY_LEVEL_FULL
+};
+
+/**
+ * Retrieves a list of names of all power supplies found. Types may vary.
+ *
+ * @param list A vector that will be initialized and filled with all
+ * power supply names found. It's required to free names pointed by
+ * it and clear the vector after usage.
+ *
+ * @return On success, it returns @c 0. On error, a negative value is returned
+ * and properties won't be set (pointers won't be valid);
+ */
+int sol_power_supply_get_list(struct sol_ptr_vector *list);
```

Comment:
after reviewing this file, to me it looks too much linux specific by using names. Go with handles, refcount them and so on.

On most small OS you're not going to have a meaningful name at all... and you'll need to do some queries that may return more than one value, so if you try to do one query at time, by name, you're doing much more effort.

on linux your implementation doesn't need to cache anything, just read the same files as methods are called.